ls1intum/Artemis
# `Discussion`: Add error handling when parsing email notification content

### Checklist
#### General


#### Server

### Motivation and Context


PR #4303, we introduced a new library for server side markdown-to-html-parsing, namely [commonmark](https://github.com/commonmark/commonmark-java).
This is a follow-up PR that further enhance the usage of commonmark-related functionality.

### Description

1. We add a url sanitizer to increase security, i.e., post content that is used as html content for generated mails (as of now, only in case of announcements) are sanitized with regard to embedded links.
2. We add an error handling as the parsing conducted by a commonmark class throws an error if the content that should be parsed (post content) is empty. in fact, afaik, this can only happen when intercepting requests as posts cannot be persisted with empty content. However, we add a general error handling to ensure a robust parsing process. If an error during parsing occurs, the mail content will be left empty.
3. I enhanced existing server tests by ensuring the a mocked method is called and added a new case to check the error handling

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Course with posts enabled


**TS 1 has to be used in order to test email related functionality.**

1. Log in to as Instructor
2. Navigate to Course Discussion
3. Create a new announcement post (context announcement selected) and add a link that tries to exploit the markdown vulnerabilities (e.g. check this [article](https://medium.com/taptuit/exploiting-xss-via-markdown-72a61e774bf8))
4. After saving, check that the created mail does include the linked text but **without** a working link, i.e. on click, nothing happens

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage





`PostService.java`: 96%

### Screenshots


Email with allowed link:
<img width="1278" alt="Bildschirmfoto 2021-11-24 um 10 52 10" src="https://user-images.githubusercontent.com/41366522/143215542-12e18ad7-cbc1-4f5b-aad0-69ece86ed7d9.png">

Email with sanitized, disallowed link (i.e., when the markdown tried to execute javascript):
<img width="1282" alt="Bildschirmfoto 2021-11-24 um 10 53 25" src="https://user-images.githubusercontent.com/41366522/143215613-f65a1f30-2f29-460a-8555-1c95aadd812e.png">
